### PR TITLE
Replace ISystemClock with TimeProvider in Kestrel

### LINKED ISSUE #13628: Consider not using the wall clock time for timeouts in Kestrel
### Is your feature request related to a problem? Please describe.

I have an embedded instrument with an RTC that's powered by a small gold cap. However, that battery is limited and so the instrument might forget the current date / time. If that happens, the system boots with a default date/time depending on when the Linux image was built.

There's a user interface that's rendered on an internal web browser connected to the localhost web server via SignalR.

Third, there's a way to set the date/time when lost, which directly sets the wall clock on the instrument (the time difference is typically O(months)).

When the user sets the date/time, the SignalR connection runs into a timeout [because timeouts are based on wall clock time](https://github.com/aspnet/AspNetCore/blob/v3.0.0-preview9.19424.4/src/SignalR/server/Core/src/HubConnectionContext.cs#L558) and my thought-to-be stable localhost connection breaks. I have included a retry logic just for that case.

### Describe the solution you'd like

Using wall clock for timeouts always runs into scenarios such as the one described above. In my opinion it would be better to base those decisions off of a monotonic and continuous time source (Stopwatch?).

### LINKED ISSUE #37862: Quarantining StreamPool_Heartbeat_ExpiredStreamRemoved
## Failing Test(s)

- Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved

## Error Message

```text
Assert.Equal() Failure
Expected: 1
Actual:   0
```

## Stacktrace

<details>

```text
 | Xunit.Sdk.EqualException: Assert.Equal() Failure
 | Expected: 1
 | Actual:   0
 |    at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:line 40
 |    at Xunit.Assert.Equal[T](T expected, T actual) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:line 24
 |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 490
 |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 507
 |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 507
 |    at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
 | --- End of stack trace from previous location ---
 |    at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
 |    at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 96
 | [1.219s] TestLifetime Information: Finished test StreamPool_Heartbeat_ExpiredStreamRemoved in 1.2182106s
```
</details>

## Logs

<details>

```text
[xUnit.net 00:00:05.44]     Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved [FAIL]
[xUnit.net 00:00:05.44]       Assert.Equal() Failure
[xUnit.net 00:00:05.44]       Expected: 1
[xUnit.net 00:00:05.44]       Actual:   0
[xUnit.net 00:00:05.44]       Stack Trace:
[xUnit.net 00:00:05.45]         /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs(490,0): at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved()
[xUnit.net 00:00:05.45]         /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs(507,0): at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved()
[xUnit.net 00:00:05.45]         /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs(507,0): at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved()
[xUnit.net 00:00:05.45]         --- End of stack trace from previous location ---
[xUnit.net 00:00:05.45]       Output:
[xUnit.net 00:00:05.45]         | [0.001s] TestLifetime Information: Starting test StreamPool_Heartbeat_ExpiredStreamRemoved at 2021-10-26T22:17:14
[xUnit.net 00:00:05.45]         | [0.023s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Connection id "0HMCOPPDQPBUI" accepted.
[xUnit.net 00:00:05.45]         | [1.164s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000000" type Bidirectional accepted.
[xUnit.net 00:00:05.45]         | [1.164s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000000" shutting down writes because: "The QUIC transport's send loop completed gracefully.".
[xUnit.net 00:00:05.45]         | [1.197s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000004" type Bidirectional accepted.
[xUnit.net 00:00:05.45]         | [1.206s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000004" write aborted by peer with error code 4294967295.
[xUnit.net 00:00:05.45]         | [1.213s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000004" shutting down writes because: "Stream aborted by peer (4294967295).".
[xUnit.net 00:00:05.45]         | [1.217s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests Error: Test threw an exception.
[xUnit.net 00:00:05.45]         | Xunit.Sdk.EqualException: Assert.Equal() Failure
[xUnit.net 00:00:05.45]         | Expected: 1
[xUnit.net 00:00:05.45]         | Actual:   0
[xUnit.net 00:00:05.45]         |    at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:line 40
[xUnit.net 00:00:05.45]         |    at Xunit.Assert.Equal[T](T expected, T actual) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:line 24
[xUnit.net 00:00:05.45]         |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 490
[xUnit.net 00:00:05.45]         |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 507
[xUnit.net 00:00:05.45]         |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 507
[xUnit.net 00:00:05.45]         |    at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
[xUnit.net 00:00:05.45]         | --- End of stack trace from previous location ---
[xUnit.net 00:00:05.45]         |    at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
[xUnit.net 00:00:05.45]         |    at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 96
[xUnit.net 00:00:05.46]         | [1.219s] TestLifetime Information: Finished test StreamPool_Heartbeat_ExpiredStreamRemoved in 1.2182106s
  Passed Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.AcceptAsync_ClientClosesConnection_ExceptionThrown [54 ms]
  Failed Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved [1 s]
  Error Message:
   Assert.Equal() Failure
Expected: 1
Actual:   0
  Stack Trace:
     at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 490
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 507
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 507
--- End of stack trace from previous location ---
  Standard Output Messages:
 | [0.001s] TestLifetime Information: Starting test StreamPool_Heartbeat_ExpiredStreamRemoved at 2021-10-26T22:17:14
 | [0.023s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Connection id "0HMCOPPDQPBUI" accepted.
 | [1.164s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000000" type Bidirectional accepted.
 | [1.164s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000000" shutting down writes because: "The QUIC transport's send loop completed gracefully.".
 | [1.197s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000004" type Bidirectional accepted.
 | [1.206s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000004" write aborted by peer with error code 4294967295.
 | [1.213s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic Debug: Stream id "0HMCOPPDQPBUI:00000004" shutting down writes because: "Stream aborted by peer (4294967295).".
 | [1.217s] Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests Error: Test threw an exception.
 | Xunit.Sdk.EqualException: Assert.Equal() Failure
 | Expected: 1
 | Actual:   0
 |    at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:line 40
 |    at Xunit.Assert.Equal[T](T expected, T actual) in C:\Dev\xunit\xunit\src\xunit.assert\Asserts\EqualityAsserts.cs:line 24
 |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 490
 |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 507
 |    at Microsoft.AspNetCore.Server.Kestrel.Transport.Quic.Tests.QuicConnectionContextTests.StreamPool_Heartbeat_ExpiredStreamRemoved() in /_/src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:line 507
 |    at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
 | --- End of stack trace from previous location ---
 |    at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
 |    at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 96
 | [1.219s] TestLifetime Information: Finished test StreamPool_Heartbeat_ExpiredStreamRemoved in 1.2182106s

```
</details>

## Build

https://dev.azure.com/dnceng/public/_build/results?buildId=1441586&view=logs&jobId=3f6d4e0f-1b71-56b5-361e-d95b6e6da15a

### LINKED ISSUE #47472: Onboard aspnetcore into the new TimeProvider type from runtime
Now that TimeProvider has been added to the platform in https://github.com/dotnet/runtime/pull/83604,  we should make the following changes in the aspnetcore layer for .NET 8:

- Remove all the copies of ISystemClock that exist in the repo (there are at least 6, with types that implement them), particularly all the ones that are not publicly exposed, and replace the places where they are used with TimeProvider.
- Two of these copies of ISystemClock are actually public and part of some of our API surface: Microsoft.AspNetCore.Authentication.ISystemClock, and Microsoft.AspNetCore.Server.kestrel.Core.Internal.Infrastructure.ISystemClock. We should consider deprecating those as well as the affected API that exposes them, and consider if there are new APIs/overloads that should be exposed that use TimeProvider as opposed to ISystemClock.

cc: @tarekgh @Tratcher @adityamandaleeka @davidfowl @mkArtakMSFT 

- [x] Authentication https://github.com/dotnet/aspnetcore/pull/47717
- [x] OutputCaching #47753
- [x] ResponseCaching #47753
- [x] YARP https://github.com/microsoft/reverse-proxy/pull/2108
- [x] SignalR Common - #47895
- [x] Kestrel - https://github.com/dotnet/aspnetcore/issues/13628 suggests using a monotonic value like SignalR.
- [x] Reconcile to a shared MockTimeProvider across AspNetCore. - https://github.com/dotnet/aspnetcore/issues/13628

### PR DESCRIPTION
**Pull request by Tratcher**
Fixes https://github.com/dotnet/aspnetcore/issues/47472
Fixes https://github.com/dotnet/aspnetcore/issues/13628

This replaces the internal ISystemClock types with the new TimeProvider abstraction. I also included https://github.com/dotnet/aspnetcore/issues/13628 which switches most operations to timestamps instead of utc dates, which may clean up the code a bit and allow for other micro-optimizations by working more directly with longs than DateTimeOffset.

Fixes #37862 (quarantined test which hasn't failed in 30+ days)

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs [hunk 1]
```diff
@@ -237,7 +237,7 @@ private void Tick()
         }
 
         // It's safe to use UtcNowUnsynchronized since Tick is called by the Heartbeat.
```
Comment (sebastienros) [inline src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs:239]: Comment needs to be updated

File: src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs [hunk 2]
```diff
@@ -720,9 +720,24 @@ private record RequestState(
         public int ActiveConcurrentConnections { get; set; }
     };
 
-    private class TestSystemClock : ISystemClock
+    private class TestTimeProvider : TimeProvider
```
Comment (sebastienros) [inline src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:723]: Should it override `CreateTimer` throwing `NotImplementedException` to ensure that no current or future code is using it as it is not mocked?

Comment (sebastienros) [reply src/Servers/Kestrel/Transport.Quic/test/QuicConnectionContextTests.cs:723]: This would be called indirectly by `Task.WaitAsync` or `Task.DelayAsync` so it could be missed.

**Review by sebastienros — APPROVED** [review]

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs [hunk 3]
```diff
@@ -953,7 +953,7 @@ public async Task DATA_Received_SlowlyDueToConnectionFlowControl_DoesNotAbortCon
 
         _mockTimeoutHandler.Verify(h => h.OnTimeout(It.IsAny<TimeoutReason>()), Times.Never);
 
-        AdvanceTime(TimeSpan.FromTicks(1));
+        AdvanceTime(TimeSpan.FromMicroseconds(1));
```
Comment (JamesNK) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:956]: nit: FromMicroseconds here vs FromMilliseconds everywhere else.

Why the change from ticks to milliseconds?

File: src/Servers/Kestrel/shared/test/Http3/Http3InMemory.cs [hunk 4]
```diff
@@ -205,20 +201,26 @@ public void AdvanceTime(TimeSpan timeSpan)
...
+        TriggerTick(endTime);
     }
 
     public void TriggerTick(DateTimeOffset now)
     {
         _mockTimeProvider.SetUtcNow(now);
+        Connection?.Tick(_mockTimeProvider.GetTimestamp());
+    }
+
+    public void TriggerTick(long now)
```
Comment (JamesNK) [inline src/Servers/Kestrel/shared/test/Http3/Http3InMemory.cs:220]: What is the long? I think variables like this need to say what the value is, e.g. nowTimestamp, or just timestamp

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs (hunk 3 above)
Comment (Tratcher) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:956]: Reverted. I was having trouble with rounding errors with the mock time provider.

**Comment by danmoseley** [comment]
will we eventually have only one TestTimeProvider? looks like 4 are copies, plus the SignalR one is possibly mergeable

**Comment by Tratcher** [comment]
> will we eventually have only one TestTimeProvider? looks like 4 are copies, plus the SignalR one is possibly mergeable

I got rid of two custom time providers in this PR, one in product code and one test. Once this review is done I can look at reconciling across AspNetCore.

File: src/Shared/test/MockTimeProvider.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    public int UtcNowCalled { get; private set; }
+
+    public void Advance(TimeSpan timeSpan)
+    {
+        Interlocked.Add(ref _utcTicks, timeSpan.Ticks);
+        Interlocked.Add(ref _timestamp, timeSpan.Ticks * (_timestampFrequency / TimeSpan.TicksPerSecond));
+    }
+
+    public void AdvanceTo(DateTimeOffset newUtcNow)
```
Comment (BrennanConroy) [inline src/Shared/test/MockTimeProvider.cs:52]: Should the AdvanceTo calls check that the time doesn't go backwards?

Comment (Tratcher) [reply src/Shared/test/MockTimeProvider.cs:52]: Backwards is fine, except if you make the timestamp go negative. We could check for that.

**Comment by sebastienros** [comment]
/benchmark plaintext aspnet-citrine-lin kestrel --application.sdkVersion 8.0.100-preview.5.23260.3

**Comment by sebastienros** [comment]
```
| application           | plaintext.base            | plaintext.pr              |        |
| --------------------- | ------------------------- | ------------------------- | ------ |
| CPU Usage (%)         |                        99 |                        99 |  0.00% |
| Cores usage (%)       |                     2,774 |                     2,777 | +0.11% |
| Working Set (MB)      |                       122 |                       121 | -0.82% |
| Private Memory (MB)   |                       655 |                       651 | -0.61% |
| Build Time (ms)       |                     4,778 |                     4,378 | -8.37% |
| Start Time (ms)       |                       210 |                       210 |  0.00% |
| Published Size (KB)   |                    96,262 |                    96,262 |  0.00% |
| Symbols Size (KB)     |                        53 |                        53 |  0.00% |
| .NET Core SDK Version | 8.0.100-preview.5.23260.3 | 8.0.100-preview.5.23260.3 |        |


| load                   | plaintext.base | plaintext.pr |         |
| ---------------------- | -------------- | ------------ | ------- |
| CPU Usage (%)          |             93 |           94 |  +1.08% |
| Cores usage (%)        |          2,613 |        2,642 |  +1.11% |
| Working Set (MB)       |             48 |           48 |   0.00% |
| Private Memory (MB)    |            370 |          370 |   0.00% |
| Start Time (ms)        |              0 |            0 |         |
| First Request (ms)     |            103 |           99 |  -3.88% |
| Requests/sec           |     11,071,054 |   11,138,157 |  +0.61% |
| Requests               |    167,173,536 |  168,100,896 |  +0.55% |
| Mean latency (ms)      |           1.28 |         1.22 |  -4.69% |
| Max latency (ms)       |          61.79 |        91.46 | +48.02% |
| Bad responses          |              0 |            0 |         |
| Socket errors          |              0 |            0 |         |
| Read throughput (MB/s) |       1,331.20 |     1,341.44 |  +0.77% |
| Latency 50th (ms)      |           0.74 |         0.73 |  -1.34% |
| Latency 75th (ms)      |           1.13 |         1.10 |  -2.65% |
| Latency 90th (ms)      |           1.92 |         1.81 |  -5.73% |
| Latency 99th (ms)      |          14.03 |        13.36 |  -4.78% |
```

**Comment by Tratcher** [comment]
/benchmark plaintext aspnet-citrine-lin kestrel --application.sdkVersion 8.0.100-preview.5.23260.3

**Comment by Tratcher** [comment]
```
| application           | plaintext.base            | plaintext.pr              |        |
| --------------------- | ------------------------- | ------------------------- | ------ |
| CPU Usage (%)         |                       100 |                        99 | -1.00% |
| Cores usage (%)       |                     2,787 |                     2,773 | -0.50% |
| Working Set (MB)      |                       124 |                       120 | -3.23% |
| Private Memory (MB)   |                       656 |                       651 | -0.76% |
| Build Time (ms)       |                     4,749 |                     4,289 | -9.69% |
| Start Time (ms)       |                       208 |                       207 | -0.48% |
| Published Size (KB)   |                    96,262 |                    96,262 |  0.00% |
| Symbols Size (KB)     |                        53 |                        53 |  0.00% |
| .NET Core SDK Version | 8.0.100-preview.5.23260.3 | 8.0.100-preview.5.23260.3 |        |


| load                   | plaintext.base | plaintext.pr |         |
| ---------------------- | -------------- | ------------ | ------- |
| CPU Usage (%)          |             94 |           95 |  +1.06% |
| Cores usage (%)        |          2,627 |        2,664 |  +1.41% |
| Working Set (MB)       |             48 |           48 |   0.00% |
| Private Memory (MB)    |            370 |          370 |   0.00% |
| Start Time (ms)        |              0 |            0 |         |
| First Request (ms)     |             92 |           96 |  +4.35% |
| Requests/sec           |     11,113,351 |   11,232,468 |  +1.07% |
| Requests               |    167,519,747 |  169,255,920 |  +1.04% |
| Mean latency (ms)      |           1.62 |         1.21 | -25.31% |
| Max latency (ms)       |         199.10 |        67.80 | -65.95% |
| Bad responses          |              0 |            0 |         |
| Socket errors          |              0 |            0 |         |
| Read throughput (MB/s) |       1,331.20 |     1,351.68 |  +1.54% |
| Latency 50th (ms)      |           0.74 |         0.73 |  -2.16% |
| Latency 75th (ms)      |           1.12 |         1.09 |  -2.68% |
| Latency 90th (ms)      |           1.88 |         1.74 |  -7.45% |
| Latency 99th (ms)      |          16.75 |        13.51 | -19.34% |
```

The few metrics that looked concerning are unstable across runs. I think this is good as far as perf goes.

File: src/Shared/test/MockTimeProvider.cs (hunk 5 above)
Comment (Tratcher) [reply src/Shared/test/MockTimeProvider.cs:57]: Right, that's why I added a check for negative _timestamp and _utcTicks.

Comment (BrennanConroy) [reply src/Shared/test/MockTimeProvider.cs:57]: That makes it worse. Because now a test will be flaky if it does any sort of time reversing.

And I still haven't seen a scenario where we do want to go backwards. So why let that happen until such a scenario occurs?

Comment (Tratcher) [reply src/Shared/test/MockTimeProvider.cs:57]: Hmmm that's a good point about flakyness. I'll see if preventing reversals causes any issues.

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        if (timeSpan == TimeSpan.MaxValue)
+        {
+            return long.MaxValue;
+        }
+        if (timeSpan == TimeSpan.MinValue)
+        {
+            return long.MinValue;
+        }
+        // The tick frequency should be equal or greater than TicksPerSecond
+        return timeSpan.Ticks * (tickFrequency / TimeSpan.TicksPerSecond);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs:22]: `long` division can lose accuracy if the tick frequency is fractionally different than TimeSpan.TicksPerSecond. [`TimeProvider.GetElapsedTime` uses `double` or the division](https://github.com/dotnet/runtime/blob/f73356c1f4b9b292dc4fb02f1dd47e4a61e423f5/src/libraries/Common/src/System/TimeProvider.cs#L113).

```suggestion
        return (long)(timeSpan.Ticks * ((double)tickFrequency / TimeSpan.TicksPerSecond));
```
It's weird to me that we special case min and max value, but otherwise let this potentially overflow. We usually call `ToTicks()` on user-provided timespans like configurable timeouts, right? That means it should be called fairly infrequently. If so, I wonder if we should use checked math to be cautious.

```suggestion
        checked
        {
            return (long)(timeSpan.Ticks * ((double)tickFrequency / TimeSpan.TicksPerSecond));
        }
```

I see `TimeProvider` doesn't bother with checked math for this. I'm not sure why, but it seems like it'd be really bad if this overflowed. We could special case `tickFrequency == TimeSpan.TicksPerSecond` to avoid the checked math in almost all real scenarios. @tarekgh

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    public static long GetTimestamp(this TimeProvider timeProvider, TimeSpan timeSpan)
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs:25]: This is only called by tests, right? Can we move this to the `MockTimeProvider`?

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        return timeProvider.GetTimestamp(timeProvider.GetTimestamp(), timeSpan);
+    }
+
+    public static long GetTimestamp(this TimeProvider timeProvider, long timeStamp, TimeSpan timeSpan)
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs:30]: Nit:

```suggestion
    private static long GetTimestamp(this TimeProvider timeProvider, long timeStamp, TimeSpan timeSpan)
```

File: src/Shared/test/MockTimeProvider.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        var nowTicks = newUtcNow.UtcTicks;
+        var priorTicks = Interlocked.Exchange(ref _utcTicks, nowTicks);
+        // Advance Timestamp by the same amount.
+        // Known timestamp frequencies are the same or larger than TicksPerSecond.
+        var timestampOffset = (nowTicks - priorTicks) * (_timestampFrequency / TimeSpan.TicksPerSecond);
+        Interlocked.Add(ref _timestamp, timestampOffset);
+
+        if (_utcTicks < 0 || _timestamp < 0)
```
Comment (halter73) [inline src/Shared/test/MockTimeProvider.cs:66]: I know this is just test code, but why not verify that `timestampOffset` is positive? Do we have a test that intentionally moves the clock backwards? I think this that's what @BrennanConroy was proposing.

```suggestion
        if (timestampOffset < 0 || _utcTicks < 0 || _timestamp < 0)
```

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs (hunk 6 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs:22]: tickFrequency only matches TimeSpan.TicksPerSecond on Windows, not Linux.

I special cased MaxValue because it's used in tests and because customers are likely to use it when they think they're disabling timeouts.

I'll harden this code path.

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs (hunk 7 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs:25]: Http2Connection also uses it.

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs (hunk 8 above)
Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeExtensions.cs:30]: ? This is used by Http2Connection and several tests.

**Comment by Tratcher** [comment]
/benchmark plaintext aspnet-citrine-lin kestrel --application.sdkVersion 8.0.100-preview.5.23260.3
